lib/docker.ts

```diff
@@ -132,69 +180,110 @@ class Docker {
-            }
-          }
-        }
-      }
+  /**
+   * Attempt to perform a static scan
+   * @param extractActions array of pattern, callbacks pairs
+   */
+  public async scanStaticalyIfNeeded(
+    extractActions: ExtractAction[],
+  ): Promise<void> {
+    const size = await this.sizeSafe();
+    if (!size || size > this.GetStaticScanSizeLimit()) {
+      return;
+    }
+    try {
+      await this.extractAndCache(extractActions);
+    } catch (error) {
+      debug(error);
+      throw error;
     }
-    return result;
   }
 
   /**
-   * Saves the docker image as a TAR file to a temporary location and extract
-   * the specified files from it.
-   * @param txtPatterns list of plain text key files paths patterns to extract and return as strings
-   * @param md5Patterns list of binary key files paths patterns to extract and return their MD5 sum
-   * @return list of plain text files and list of MD5 sums
+   * Get file product that was previously retrieved and cached or by using
+   *  runtime method after applying the specified callback
+   * @param filename name of file to retrieve its associated string product
+   * @param searchActionName name of a search action
+   * @param callbacks optional array of callbacks to call when runtime method
+   *  is used to retrieve the file
+   * @returns map of file products by callback name
    */
-  public async extract(
-    txtPatterns: string[],
-    md5Patterns: string[] = [],
-  ): Promise<ExtractedKeyFiles> {
-    return this.save(async (err, imageTarPath) => {
-      if (err) {
-        throw err;
+  public async getFileProduct(
+    filename: string,
+    searchActionName: string,
+    extractCallback?: ExtractCallback,
+  ): Promise<string | Buffer> {
+    if (Reflect.has(this.extractProductsByFilename, filename)) {
+      const callbackProducts = this.extractProductsByFilename[filename];
+      if (Reflect.has(callbackProducts, searchActionName)) {
+        return callbackProducts[searchActionName];
       }
-      return await this.analyze(imageTarPath, txtPatterns, md5Patterns);
-    });
+    }
+    const content = (await this.catSafe(filename)).stdout;
+    return extractCallback
+      ? extractCallback(Buffer.from(content, "utf8"))
+      : content;
+  }
+
+  /**
+   * Retrieve all filenames with products of the specified action name
+   * @param searchActionName name of a search action
+   */
+  public getFileProducts(searchActionName: string = "txt"): FilenameProducts {
+    return Object.assign(
+      {},
+      // go over all filenames
+      ...Object.keys(this.extractProductsByFilename)
+        // check for product associated with the specified action name
+        .filter((filename) =>
+          Reflect.has(
+            this.extractProductsByFilename[filename],
+            searchActionName,
+          ),
+        )
+        // create a map of this filename and product
+        .map((filename) => ({
+          [filename]: this.extractProductsByFilename[searchActionName],
```

Comment:
Sorry, but this will not work. The line should be:
```suggestion
          [filename]: this.extractProductsByFilename[filename][searchActionName]
```

Reply:
@mladkau great catch updated the test to verify the products of the file and not just the filename existence